### PR TITLE
Refactor of distcp to allow for more complex publish targets and easi…

### PR DESCRIPTION
**Pull request by ibuenros**
…er DatasetFinder implementation

### CODE REVIEW

**Comment by ibuenros** [comment]
@liyinan926 can you review?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/CopyableDatasetMetadata.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+@EqualsAndHashCode(callSuper=false)
+public class CopyableDatasetMetadata {
+
+  public CopyableDatasetMetadata(CopyableDataset copyableDataset, Path datasetTargetRoot) {
+    this.datasetRoot = copyableDataset.datasetRoot();
+    this.datasetTargetRoot = datasetTargetRoot;
+  }
+
+  @Getter
+  private Path datasetRoot;
```
Comment (liyinan926) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/CopyableDatasetMetadata.java]: Seems they can be final.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/CopyableDatasetMetadata.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  @Getter
+  private Path datasetTargetRoot;
+  private static Gson gson = new Gson();
```
Comment (liyinan926) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/CopyableDatasetMetadata.java]: Why not make it a `static final Gson GSON = ...`?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/RecursiveCopyableDataset.java [hunk 3]
```diff
@@ -44,20 +44,14 @@
...
   private final Path rootPath;
   private final FileSystem fs;
-  private final Path targetDirectory;
   private final Properties properties;
   private LoadingCache<Path, OwnerAndPermission> ownerAndPermissionCache;
   private final PathFilter pathFilter;
 
   public RecursiveCopyableDataset(final FileSystem fs, Path rootPath, Properties properties) {
 
-    Preconditions.checkArgument(
```
Comment (liyinan926) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/RecursiveCopyableDataset.java]: Why this is gone?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/CopySource.java [hunk 4]
```diff
@@ -129,6 +137,19 @@ protected FileSystem getSourceFileSystem(State state) throws IOException {
     return FileSystem.get(URI.create(uri), conf);
   }
 
+  private FileSystem getTargetFileSystem(State state) throws IOException {
+    return WriterUtils.getWriterFS(state, 1, 0);
+  }
+
+  private Path getTargetRoot(State state, DatasetFinder<?> datasetFinder, CopyableDataset dataset) {
+    Preconditions.checkArgument(state.contains(ConfigurationKeys.DATA_PUBLISHER_FINAL_DIR));
```
Comment (liyinan926) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/CopySource.java]: Better to use the second argument to include a message here.

File: gobblin-data-management/src/main/java/gobblin/data/management/util/PathUtils.java [hunk 5]
```diff
@@ -53,4 +57,16 @@ public static Path getPathWithoutSchemeAndAuthority(Path path) {
...
     return new Path(StringUtils.removeStart(path.toString(), Path.SEPARATOR));
   }
+
+  /**
+   * Finds the deepest ancestor of input that is not a glob.
+   */
+  public static Path deepestNonGlobPath(Path input) {
+    Path commonRoot = input;
+
+    while(GLOB_TOKENS.matcher(commonRoot.toString()).find()) {
```
Comment (liyinan926) [inline gobblin-data-management/src/main/java/gobblin/data/management/util/PathUtils.java]: Seems you can also do `GLOB_TOKENS.matcher(commonRoot.getName()).matches()`?

Comment (liyinan926) [inline gobblin-data-management/src/main/java/gobblin/data/management/util/PathUtils.java]: Also the condition is better to be `commonRoot != null && ...`.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/RecursiveCopyableDataset.java (hunk 3 above)
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/RecursiveCopyableDataset.java]: It is no longer the responsibility of the Dataset to figure out the target, so it doesn't need this option anymore.

File: gobblin-data-management/src/main/java/gobblin/data/management/util/PathUtils.java (hunk 5 above)
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/util/PathUtils.java]: For the first comment, no, we need to match against entire string. For example `/foo/*/bar` is a glob, even though the name doesn't include a glob. I will fix the second.

**Comment by ibuenros** [comment]
@liyinan926 addressed comments.

**Comment by liyinan926** [comment]
LGTM.

**Pull request merged**